wsagent/che-wsagent-core/src/main/java/org/eclipse/che/wsagent/server/CheWsAgentCorsAllowedOriginsProvider.java

```diff
@@ -0,0 +1,99 @@
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.wsagent.server;
+
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+import java.io.IOException;
+import javax.inject.Named;
+import javax.ws.rs.core.UriBuilder;
+import org.eclipse.che.api.core.ApiException;
+import org.eclipse.che.api.core.model.workspace.runtime.Machine;
+import org.eclipse.che.api.core.model.workspace.runtime.Server;
+import org.eclipse.che.api.core.rest.HttpJsonRequestFactory;
+import org.eclipse.che.api.workspace.server.WorkspaceService;
+import org.eclipse.che.api.workspace.shared.dto.WorkspaceDto;
+import org.eclipse.che.commons.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provider of "cors.allowed.origins" setting for CORS Filter of WS Agent. Provides the value such
+ * algorithm:
+ *
+ * <ul>
+ *   <li>If set che.wsagent.cors.allowed_origins
+ *   <li>Server with "ide" attribute in workspace config
+ *   <li>Server from url of "ide" link in workspace config
+ *   <li>che.api
+ * </ul>
+ */
+public class CheWsAgentCorsAllowedOriginsProvider implements Provider<String> {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(CheWsAgentCorsAllowedOriginsProvider.class);
+  private final String allowedOrigins;
+  private final String apiEndpoint;
+
+  @Inject
+  public CheWsAgentCorsAllowedOriginsProvider(
+      @Named("che.api") String apiEndpoint,
+      @Nullable @Named("che.wsagent.cors.allowed_origins") String allowedOrigins,
+      HttpJsonRequestFactory httpJsonRequestFactory)
+      throws ApiException, IOException {
+    this.apiEndpoint = apiEndpoint;
+
+    if (allowedOrigins == null) {
+      final UriBuilder builder =
+          UriBuilder.fromUri(apiEndpoint)
+              .path(WorkspaceService.class)
+              .path(WorkspaceService.class, "getByKey")
+              .queryParam("includeInternalServers", "true");
+
+      String href = builder.build(System.getenv("CHE_WORKSPACE_ID")).toString();
+      WorkspaceDto workspaceDto =
+          httpJsonRequestFactory.fromUrl(href).useGetMethod().request().asDto(WorkspaceDto.class);
+
+      String ideUrl = getIdeUrl(workspaceDto);
+      if (ideUrl != null) {
+        this.allowedOrigins = UriBuilder.fromUri(ideUrl).replacePath(null).build().toString();
+      } else {
+        this.allowedOrigins =
+            UriBuilder.fromUri(workspaceDto.getLinks().getOrDefault("ide", apiEndpoint))
+                .replacePath(null)
+                .build()
+                .toString();
+      }
+
+    } else {
+      this.allowedOrigins = allowedOrigins;
+    }
+  }
+
+  @Override
+  public String get() {
+    LOG.debug("allowedOrigins {} ", allowedOrigins);
+    return allowedOrigins;
+  }
+
+  private String getIdeUrl(WorkspaceDto workspaceDto) {
```

Comment:
Maybe it's  easier to just get the link by (this | https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-workspace-shared/src/main/java/org/eclipse/che/api/workspace/shared/Constants.java#L24 ) rel from WS links ?

Reply:
I'm using this link here https://github.com/eclipse/che/pull/12404/files#diff-11cbb1d2057d9e410a50d3a7c532cc2bR71 in case if there is no server declared with ide attribute aka Che6 mode for GWT Ide

Reply:
So can it be different between server url and link in workspace?

Reply:
yes